lib/filterx/expr-function.c

```diff
@@ -115,23 +136,53 @@ static void
 _simple_free(FilterXExpr *s)
 {
   FilterXSimpleFunction *self = (FilterXSimpleFunction *) s;
-  filterx_function_args_free(self->args);
+  g_ptr_array_free(self->args, TRUE);
   filterx_function_free_method(&self->super);
 }
 
+static inline gboolean
+_simple_process_args(FilterXSimpleFunction *self, FilterXFunctionArgs *args, GError **error)
+{
+  guint64 len = filterx_function_args_len(args);
+
+  for (guint64 i = 0; i < len; i++)
+    {
+      FilterXExpr *expr = filterx_function_args_get_expr(args, i);
+      if (!expr)
+        {
+          g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_UNEXPECTED_ARGS, "can't get argument");
+          return FALSE;
+        }
+
+      g_ptr_array_add(self->args, expr);
+    }
+
+  return TRUE;
+}
+
 FilterXExpr *
 filterx_simple_function_new(const gchar *function_name, FilterXFunctionArgs *args,
-                            FilterXSimpleFunctionProto function_proto)
+                            FilterXSimpleFunctionProto function_proto, GError **error)
 {
   FilterXSimpleFunction *self = g_new0(FilterXSimpleFunction, 1);
 
   filterx_function_init_instance(&self->super, function_name);
   self->super.super.eval = _simple_eval;
   self->super.super.free_fn = _simple_free;
-  self->args = args;
   self->function_proto = function_proto;
 
+  self->args = g_ptr_array_new_full(8, (GDestroyNotify) filterx_expr_unref);
```

Comment:
this 8 seems a little bit ad hoc to me, especially you can have the number of arguments there

Reply:
It's based on this:
`self->positional_args = g_ptr_array_new_full(8, (GDestroyNotify) filterx_function_arg_free);`

8 is just the initial capacity here.

Reply:
I understood. Fixed.